stablehlo/reference/Ops.cpp

```diff
@@ -224,12 +236,187 @@ Tensor evalTransposeOp(const Tensor &operand, ArrayRef<int64_t> permutation,
   return result;
 }
 
+SmallVector<Tensor> evalWhileOp(ArrayRef<Tensor> operand, Region &cond,
+                                Region &body, const Scope &scope) {
+  SmallVector<Tensor> runtimeResults(operand);
+
+  auto runtimeCondResultsOrErr = evalRegion(cond, operand, &scope);
+  if (!runtimeCondResultsOrErr || runtimeCondResultsOrErr->size() != 1)
+    llvm::report_fatal_error("Failed to evaluate cond");
+
+  auto runtimeCondResult = (*runtimeCondResultsOrErr)[0];
+  while (runtimeCondResult.get(*runtimeCondResult.index_begin())
+             .getBooleanValue()) {
+    auto runtimeBodyResultsOrErr = evalRegion(body, runtimeResults, &scope);
+    if (!runtimeBodyResultsOrErr)
+      llvm::report_fatal_error("Failed to evaluate body");
+
+    runtimeResults = *runtimeBodyResultsOrErr;
+    runtimeCondResultsOrErr = evalRegion(cond, runtimeResults, &scope);
+    if (!runtimeCondResultsOrErr || runtimeCondResultsOrErr->size() != 1)
+      llvm::report_fatal_error("Failed to evaluate cond");
+
+    runtimeCondResult = (*runtimeCondResultsOrErr)[0];
+  }
+  return runtimeResults;
+}
+
 Tensor evalXorOp(const Tensor &lhs, const Tensor &rhs, Type resultType) {
   Tensor result(resultType);
   for (auto it = lhs.index_begin(); it != lhs.index_end(); ++it)
     result.set(*it, lhs.get(*it) ^ rhs.get(*it));
   return result;
 }
 
+llvm::Expected<SmallVector<Tensor>> evalRegion(Region &region,
+                                               ArrayRef<Tensor> args,
+                                               const Scope *const parentScope) {
+  if (!region.hasOneBlock())
+    return invalidArgument("Expected single block region");
+
+  Block &block = region.front();
+  if (block.getArguments().size() != args.size())
+    report_fatal_error(invalidArgument(
+        "Expected same amount of block arguments and runtime arguments (%d)",
+        args.size()));
+
+  Scope scope(parentScope);
+  for (auto [ssaArg, runtimeArg] : llvm::zip(block.getArguments(), args)) {
+    assert(ssaArg.getType() == runtimeArg.getType());
+    scope.add(ssaArg, runtimeArg);
+  }
+
+  for (Operation &op : block) {
+    SmallVector<Value> results(op.getResults().begin(), op.getResults().end());
```

Comment:
Can we avoid the need to create a temporary vector? The `add` overload doesn't necessarily have to an `ArrayRef`. Something zippable would do as well.

Reply:
done. Used `ValueRange` instead.